### ISSUE TITLE
Shipment email broken when creating shipment with multiple tracks

### ISSUE THREAD

**Issue by mam08ixo** [magento/magento2#20343]
### Preconditions
1. Magento CE 2.2.8-dev (at least 4b3b669) or Magento CE 2.3-develop (at least 2307e16) is installed with sample data.
2. Email sending is properly set up.
3. The configuration setting _Sales → Sales Emails → General Settings → Asynchronous sending_ is set to _Disable_ (default).
4. At least one item exists in the `sales_shipment_track` database table.

### Steps to reproduce
1. Create a new order. Any (non-virtual) items, shipping method, payment method will work to reproduce the issue.
2. In admin panel, open the new order and click _Ship_ in the upper page main actions.
3. On the _New Shipment_ page
    - add multiple tracking numbers in the _Shipping Information_ section
    - check _ Email Copy of Shipment_ in the _Shipment Options_ section
4. Submit Shipment

### Expected result

1. The _Shipment Confirmation Email_ contains shipment details, including the tracking numbers.

### Actual result
1. The _Shipment Confirmation Email_ contains nothing but the error message 
    > Error filtering template: Item (Magento\Sales\Model\Order\Shipment\Track) with the same ID "1" already exists. 

### Further Information

**tl;dr**:
When creating a new shipment, the new tracks are added to the shipment's tracks collection (indices: 0, 1, 2, …) but the collection is not marked as loaded. When the email template then iterates over the tracks, the collection gets loaded from the database (indices: 1, 2, 3, …). As there already is an item with the index "1" in the collection, an exception is thrown.

**full story**:
The common entrypoint is [`\Magento\Sales\Model\Order\ShipmentDocumentFactory::create`](https://github.com/magento/magento2/blob/4b3b6699a6e1d3f70ad1c6006583110bf7987330/app/code/Magento/Sales/Model/Order/ShipmentDocumentFactory.php#L95). Each track submitted during the UI or REST action is added to the shipment here. In [`\Magento\Sales\Model\Order\Shipment::addTrack`](https://github.com/magento/magento2/blob/4b3b6699a6e1d3f70ad1c6006583110bf7987330/app/code/Magento/Sales/Model/Order/Shipment.php#L400), the new track is added to a [new collection, created using a factory](https://github.com/magento/magento2/blob/4b3b6699a6e1d3f70ad1c6006583110bf7987330/app/code/Magento/Sales/Model/Order/Shipment.php#L346). After the tracks have been added, the track collection's `_isCollectionLoaded` flag is `null`.

When the email sending process kicks in, the [track template iterates over the shipment's tracks](https://github.com/magento/magento2/blob/4b3b6699a6e1d3f70ad1c6006583110bf7987330/app/code/Magento/Sales/view/frontend/templates/email/shipment/track.phtml#L12). [`\Magento\Sales\Model\Order\Shipment::getAllTracks`](https://github.com/magento/magento2/blob/4b3b6699a6e1d3f70ad1c6006583110bf7987330/app/code/Magento/Sales/Model/Order/Shipment.php#L366) attempts to load the tracks via the collection (which already contains the new tracks with numeric indices starting from 0). As the collection was not marked loaded, all tracks are queried from the database (which contains rows with numeric primary keys starting from 1). As soon as the first [item from the database gets added to the collection](\Magento\Framework\Data\Collection\AbstractDb::loadWithFilter), an [exception is thrown](https://github.com/magento/magento2/blob/4b3b6699a6e1d3f70ad1c6006583110bf7987330/lib/internal/Magento/Framework/Data/Collection.php#L404) because there is already an item with index "1" in the collection.

**regression**:
This all worked in M2.2.7 and lower because the query which loads all tracks from the database during email sending never returned any results. The tracks collection of new (not yet persisted) shipments was [initialized with an empty shipment filter](https://github.com/magento/magento2/blob/2.2.7/app/code/Magento/Sales/Model/Order/Shipment.php#L346), resulting in the following query:

```sql
SELECT `main_table`.* FROM `sales_shipment_track` AS `main_table` WHERE (`parent_id` = '')
```

With the commit mentioned in the beginning, this was changed. The tracks collection of new (not yet persisted) shipments has [no filters added at all](https://github.com/magento/magento2/blob/4b3b6699a6e1d3f70ad1c6006583110bf7987330/app/code/Magento/Sales/Model/Order/Shipment.php#L348-L349), now resulting in the following query:

```sql
SELECT `main_table`.* FROM `sales_shipment_track` AS `main_table`
```

This loads all tracks from the database, including the "duplicate" index.

**Comment by magento-engcom-team** [comment]
Hi @mam08ixo. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@mam08ixo do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
Hi @Surabhi-Cedcoss. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @mam08ixo thank you for you report, this issue has already fixed in 2.3-develop branch by @Nazar65  
https://github.com/magento/magento2/pull/19981

**Comment by mam08ixo** [comment]
@engcom-backlog-nazar Good to know. Now I ran into this issue on 2.2-develop which does contain the bug but not the fix.

**Comment by ghost** [comment]
@mam08ixo https://github.com/magento/magento2/pull/20184 will be merged soon

**Comment by mam08ixo** [comment]
@engcom-backlog-nazar Thank you! I can confirm that this issue no longer occurs when above PR is applied.
